### PR TITLE
FEATURE: Always track clicks using AJAX.

### PR DESCRIPTION
**Pull request by nbianca**

### CODE REVIEW

**Comment by davidtaylorhq** [comment]
This is great - I didn't realise we already had some AJAX tracking! The biggest risk with the click tracking would be if we accidentally reset the counters. But almost all the changes here are client-side, so there's no risk of that 😄 

I tested it out locally and it worked sometimes, but other times the tracking would be missed. This was more common when using the site as an anonymous user.

The network requests were looking like this:

Screenshot 2019-04-15 at 13 46 58

So the browser was starting to load the destination site before it had loaded the CSRF token or the `track` endpoint. Therefore the browser cancelled the pending requests, and the click was never tracked.

There are a couple of things from that:
1. Small thing: The CSRF token request will add some extra overhead for anonymous users. If we want to speed things up, we could change the `track` endpoint to `GET`. That would make spoofing clicks slightly easier, but we already have user/IP rate limiters. Maybe @eviltrout can decide on this?

2. Bigger problem: The browser is going ahead and loading the new page before the track request has been sent. We need to intercept the event, stop it, then trigger the navigation later. All the code for that is there **but** we're watching the wrong event. All the tracking currently lives on the `mouseup` event, rather than the `click` event. Intercepting the `mouseup` event doesn't prevent the browser navigation. This can be fixed by doing a find/replace to convert all instances of `mouseup.discourse-redirect` to `click.discourse-redirect`. Then the network graph looks like this:

    Screenshot 2019-04-15 at 13 43 55

   I think switching to the `click` event could break the middle-click tracking, so it will need some more work. It would also be cool if we can add tracking for <kbd>Ctrl/Cmd</kbd><kbd>Click</kbd> events.

File: app/assets/javascripts/discourse/lib/click-track.js.es6 [hunk 1]
```diff
@@ -19,33 +34,33 @@ export default {
     }
 
     // cancel click if triggered as part of selection.
+    // TODO: is useless?
```
[the remark this answers is not in the document]
Comment (nbianca) [reply app/assets/javascripts/discourse/lib/click-track.js.es6]: No , I could not reproduce this and now I understand why. Everywhere we call `clickTrack` we have this little block of code that handles the exact same case:

```javascript
          const selection = window.getSelection && window.getSelection();
          if (selection.type === "Range" || selection.rangeCount > 0) {
            if (selectedText() !== "") {
              return true;
            }
          }
```

**Comment by nbianca** [comment]
@davidtaylorhq I have tried reproducing the issue you mentioned and you are right, it happens more often on Chrome with anonymous users. I could not reproduce it at all on Firefox and barely did it on Chrome when logged in.

What I also find strange is that it sometimes works and I know it's not just good timing, because `preventDefault` from this handler cancels a request and then a new request is issued by `redirectTo` called inside the promise:

Screenshot 2019-04-16 at 13 43 27

Is the browser issuing different events or in different order (mouseup & mousedown / click)? I see how registering the event handler for `click` instead of `mouseup` would help and it makes me wonder how it worked before.

The concern about the PUT request is valid and it makes sense to change it to a GET.

**Comment by davidtaylorhq** [comment]
> preventDefault from this handler cancels a request

In my testing it didn't seem like preventDefault was cancelling the request. If I comment out the `redirectTo`, the browser would still navigate to the destination. This was macOS Chrome, maybe it's different in other browsers.

> how it worked before

I think it is a race condition. Before, there was no AJAX call made, so the event handler would run super quickly and redirect somewhere else. Now, the AJAX call is quite slow, so the browser navigation wins the race.

**Comment by nbianca** [comment]
> there was no AJAX call made

We had AJAX tracking in some cases before too.

Anyway, it looks like browser navigation is handled during the `click` event and preventing default on `mouseup` / `mousedown` does nothing.

**Comment by SamSaffron** [comment]
@davidtaylorhq / @nbianca can you have a quick chat when you are both online, if you are both comfortable with the change, merge in first thing when you start your day.

**Comment by SamSaffron** [comment]
A longer term solution here is to ship the tracking to the service worker, that way there is zero chance it will be "too slow" 

I am curious though ... if google are following this pattern how come they always work?

**Comment by davidtaylorhq** [comment]
> if google are following this pattern how come they always work?

@SamSaffron Which pattern are you referring to here? I think google use the same pattern implemented in this PR, but they use Beacon requests which speed things up slightly. (And they gracefully fallback all the way to the old intermediate page system for non-js clients)

**Comment by SamSaffron** [comment]
Hmmm ... so why are we not using `Navigator.sendBeacon()` seems trivial to check for it and use it when available then fallback to full on ajax if `Navigator.sendBeacon` is missing?

**Comment by nbianca** [comment]
I will follow up with an implementation for Beacon API.

**Pull request merged**